### PR TITLE
NAS-125618 / 24.04 / Allow users to set own passwords

### PR DESCRIPTION
**Pull request by anodos325**
This commit adds a new API endpoint user.set_password that may be used to change the password of a given user.

NOTE: when authenticated session has less than FULL_ADMIN role, password changes will be rejected if the payload does not match the currently-authenticated user.

API keys granting access to this endpoint will be able to reset password of any user.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/account.py [hunk 1]
```diff
@@ -1445,6 +1448,132 @@ def update_sshpubkey(self, homedir, user, group):
...
+        if entry['locked']:
+            verrors.add(
+                'user.reset_password.username',
+                f'{username}: user account is locked.'
+            )
+
+        new_hash = crypted_password(password)
+        verrors.check()
+
+        if entry['smb']:
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/account.py:1556]: Code repetition for `__set_password`

File: src/middlewared/middlewared/plugins/account.py [hunk 2]
```diff
@@ -1445,6 +1448,132 @@ def update_sshpubkey(self, homedir, user, group):
...
+        Str('username', required=True),
+        Password('old_password', default=''),
+        Password('new_password', required=True),
+        Dict(
+            'options',
+            Bool('skip_password_check', default=False)
+        )
+    ))
+    @pass_app(require=True)
+    async def password_reset(self, app, data):
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/account.py:1463]: Why `reset` if we do `set`?

File: src/middlewared/middlewared/plugins/account.py [hunk 3]
```diff
@@ -1445,6 +1448,132 @@ def update_sshpubkey(self, homedir, user, group):
...
 
+    @no_authz_required
+    @accepts(Dict(
+        'password_reset_data',
+        Str('username', required=True),
+        Password('old_password', default=''),
+        Password('new_password', required=True),
+        Dict(
+            'options',
+            Bool('skip_password_check', default=False)
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/account.py:1459]: We can have a simpler API interface: make `old_password` not required on schema level, but check that it is set if we are not a full admin. Then this flag won't be necessary.

File: src/middlewared/middlewared/plugins/account.py (hunk 2 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/account.py]: I'm pretty bad at naming things.

**Pull request merged**